specs/plugins.md

```diff
@@ -0,0 +1,134 @@
+The plugin system will leverage `AssemblyLoadContext` from `System.Runtime.Loader` to allow for isolated loading of plugins and their dependencies. The approach will be based upon https://learn.microsoft.com/en-us/dotnet/core/tutorials/creating-app-with-plugin-support
+
+Should a the proxy be unable to load a plugin based upon the supplied configuration the proxy will exit with an error informing the user that the plugin was not able to be found.
+
+Plugins will utilize event listeners to provide a decoupled and late bound integration between the core proxy logic and the logic that the plugins provide.
+
+## Configuration file
+
+The existing `appsettings.json` file will be used to provide configuration. Both for determining which plugins are loaded into the proxy and providing configuration for each of the loaded plugins. The set of plugins to be loaded will be specified as an array of objects at the root of the configuration file in the `plugins` property. Each plugin must supply a `pluginPath` and `name` and may optionally provide `configSection` and `urlsToWatch` in the plugins array.
+- `pluginPath` - required, a string which is the relative path to the assembly containing the class definition for the plugin which will be loaded using reflection.
+- `name` - required, a string identifier for the plugin class, allows developer to ship multiple plugins in a single assembly.
+- `configSection` - optional, a string for which there must be a matching object property at the root of the configuration file. If a plugin supplies a `configSection` and no corresponding property exists in the configuration file an error will be thrown during start up of the proxy.
+- `urlsToWatch` - optional, an array of strings defining the urls for which this plugin will watch instead of the `urlsToWatch` at the root of the config file, this behavior is defined in the [Multi URL support spec](./multi-url-support.md).
+
+```json
+  "plugins": [
+    {
+        "configSection": "randomErrors",
+        "pluginPath": "RandomErrorPlugin\\msgraph-proxy-handler.dll",
+        "name": "RandomErrorPlugin",
+        "urlsToWatch": [
+            "https://graph.microsoft.com/v1.0/*",
+            "https://graph.microsoft.com/beta/*"
+        ],
+    }
+  ],
+  "randomErrors": {
+    "rate": 50,
+    "allowedErrors": [ 429, 500, 502, 503, 504, 507 ]
+  }
+```
+
+Plugins will executed in the order that they are listed in the plugins configuration array. At least one plugin must be present in the configuration, otherwise the proxy will exit with an error informing the user that their configuration is invalid and contains no plugins.
+
+## Plugins
+
+Plugin classes must implement the `IProxyPlugin` interface and provide a parameterless constructor.
+
+```cs
+public interface IProxyPlugin {
+    string Name { get; }
+    void Register(IPluginEvents pluginEvents, IConfigurationSection? configSection);
+}
+
+public interface IPluginEvents {
+    event EventHandler<InitArgs> Init;
```

Comment:
I believe it will https://learn.microsoft.com/en-us/dotnet/csharp/event-pattern#event-delegate-signatures 
The key part being:
> Events are based on delegates and are multicast delegates. That supports multiple subscribers for any event source

Plugin implementers should be able to use `pluginEvents.Init += OnInit` to register their event sinks

Reply:
Confirmed as working via a spike implementation